MobiusCore/Source/SequentialAccessGuard.swift

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2019 Spotify AB.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+
+/// Utility to catch invalid concurrent access to non-thread-safe code.
+///
+/// Like a mutex, `SequentialAccessGuard` guards a critical region. However, instead of blocking if two threads attemt
+/// to enter the critical region at once, it crashes in debug builds. In release builds, it has no effect (and also no
+/// overhead, as long as it’s only used within one module).
+///
+/// Copies of a `SequentialAccessGuard` share the same underlying mutex, and hence their critical region is the union
+/// of the critical regions of all copies.
+struct SequentialAccessGuard {
```

Comment:
I think this class provides weaker guarantees than we need. It verifies that there is no concurrent access, but what we need is that access is always performed by the same thread (or we need to ensure visibility of changes made by one thread when another thread arrives to read/write them).

Reply:
It can safely be used from multiple threads as long as there are appropriate fences in place, which will be the case if it’s used on a background dispatch queue, for example.

It’s possible to lock it to a specific queue, but that’s a much _stronger_ guarantee than we actually need. Non-overlapping-but-unfenced access from multiple threads isn’t something we can guard against specifically here, and doesn’t feel like a particularly big risk with typical iOS coding patterns. I’d rather promote appropriate tooling (Thread Sanitizer) for that.

Reply:
So all updates made on dispatch queues introduce memory barriers? What about reads from the main thread (both need some kind of fences)? It would be great if you could point me to some documentation, because I've been looking for this in order to learn more about the concurrency/memory model on iOS and failed.

Reply:
As I understand the current concurrency model for Swift, there are barriers around dispatch blocks that are “usually good enough” for “most purposes”, and everything relies on them heavily. But nothing very clearly documented. (“Most purposes” should include reads/writes in consecutive blocks on the same serial queue, if they happen to run on different threads in the thread pool.)

Basically, we should wait a few years before using Swift in production. 😢

Reply:
Two thoughts on further reflection about this: first, I think a more accurate name for this class is `ConcurrentAccessDetector`. Second, I think that since the `MobiusLoop` is not thread-safe, the condition we should check is that it is only ever invoked on a single thread. If so, this should be a `MultiThreadedAccessDetector`, and could verify that it's only ever invoked on one particular thread (`Thread.current` would seem to be a good way of handling that - and `Thread` is `Equatable`).